packages/server/src/user/__tests__/user.service.spec.ts

```diff
@@ -0,0 +1,328 @@
+
+    const userCreated = await userService.createUser(tempUserInput);
+
+    expect(userCreated.projectId).toEqual(randomProject.id);
+    expect(userCreated.username).toBe(null);
+    expect(userCreated.email).toEqual(tempUserInput.email);
+    expect(await bcrypt.compare(tempUserPasseword, userCreated.password)).toBe(true);
+    expect(userCreated.role).toBe(0);
+    expect(isEqual(userCreated.createdAt, userCreated.updatedAt)).toBe(true);
+    expect(isPast(userCreated.createdAt)).toBe(true);
+    expect(isBefore(createDate, userCreated.createdAt)).toBe(true);
+
+    await prisma.user.delete({ where: { id: userCreated.id } });
+  });
+
+  it('Create duplicated user should throw an erorr', async () => {
+    const tempUserInput = { projectId: randomUser.projectId, username: randomUser.username, email: randomUser.email, password: 'password' };
+    expect(userService.createUser(tempUserInput)).rejects.toThrow(Error);
+  });
+
+  /**
+   * Test cases for `findAllUsers()` function
+   */
+  it('Find all users should return all users regardless of project', async () => {
+    const responseUsers = await userService.findAllUsers();
+    for (const user of dummyAdmins.concat(dummyUsers)) {
+      expect(responseUsers).toContainEqual(user);
+    }
+  });
+
+  /**
+   * Test cases for `findUserByUsername()` function
+   */
+  it('Return requested user given valid username', async () => {
+    const usersWithUsername = dummyAdmins.concat(dummyUsers).filter((user) => user.username);
+    const randomUserWithUsername = usersWithUsername[Math.floor(Math.random() * usersWithUsername.length)];
+
+    const responseUser = await userService.findUserByUsername(randomUserWithUsername.projectId, randomUserWithUsername.username);
+
+    expect(responseUser).toEqual(randomUserWithUsername);
+  });
+
+  it('Return null given valid project id but different username', async () => {
+    const usersFromDiffProjects = dummyAdmins.concat(dummyUsers).filter((user) => user.projectId !== randomProject.id);
+    const randomUsername = usersFromDiffProjects[Math.floor(Math.random() * usersFromDiffProjects.length)].username;
+
+    expect(userService.findUserByUsername(randomProject.id, randomUsername)).resolves.toBe(null);
+  });
+
+  it('Return null given valid username but different project id', async () => {
+    const randomProjectId = dummyProjects.filter((project) => project.id !== randomUser.projectId)[0].id;
+    expect(userService.findUserByUsername(randomProjectId, randomUser.username)).resolves.toBe(null);
+  });
+
+  it('Return null given null username for projects that require username', async () => {
+    const usersWithUsername = dummyAdmins.concat(dummyUsers).filter((user) => user.username);
+    expect(userService.findUserByUsername(usersWithUsername[0].projectId, null)).resolves.toBe(null);
+  });
+
+  /**
+   * Test cases for `findUserByEmail()` function
```

Comment:
One thing you can do is have another describe for the functions, so

```
describe('UserService', () => 
    // All of the before each logic
   
   describe('findUserByEmail()', () => {
   
   }):
}):
```

Not something you have to change to using, but I've found it helpful for organizing tests based on methods

Reply:
I thought about that, but it seems like the `describe()` inside does not run after `beforeAll()` and `beforeEach()` inside the upper-level `describe()`.

Reply:
Jest will make sure they run in order
https://jestjs.io/docs/setup-teardown